File: backend/app/models/budget.py
import uuid
from datetime import datetime, timezone
from app.extensions import db

class Budget(db.Model):
    __tablename__ = "budgets"

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = db.Column(db.String(36), db.ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    category_id = db.Column(db.String(36), db.ForeignKey("categories.id", ondelete="CASCADE"), nullable=False, index=True)
    amount = db.Column(db.Numeric(12, 2), nullable=False)
    month = db.Column(db.SmallInteger, nullable=False)
    year = db.Column(db.SmallInteger, nullable=False)
    created_at = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at = db.Column(db.DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc), nullable=False)

    __table_args__ = (
        db.UniqueConstraint("user_id", "category_id", "month", "year", name="uq_user_category_month_year"),
    )

    # Relationships
    user = db.relationship("User", back_populates="budgets")
    category = db.relationship("Category", back_populates="budgets")

    def to_dict(self):
        return {
            "id": self.id,
            "user_id": self.user_id,
            "category_id": self.category_id,
            "category_name": self.category.name if self.category else "Unknown Category",
            "category_icon": self.category.icon if self.category else "Tag",
            "category_color": self.category.color if self.category else "#64748B",
            "amount": float(self.amount),
            "month": self.month,
            "year": self.year,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: frontend/src/types/budget.ts
export interface Budget {
  id: string;
  user_id: string;
  category_id: string;
  category_name: string;
  category_icon: string;
  category_color: string;
  amount: number;
  month: number;
  year: number;
  spent_amount: number;
  remaining_amount: number;
  percentage_used: number;
  is_exceeded: boolean;
  over_amount: number;
  created_at: string;
  updated_at?: string;
}

export interface BudgetSummary {
  total_budgeted: number;
  total_spent: number;
  total_remaining: number;
  overall_percentage_used: number;
  is_overall_exceeded: boolean;
}

export interface BudgetListResponse {
  month: number;
  year: number;
  summary: BudgetSummary;
  budgets: Budget[];
}

export interface SaveBudgetPayload {
  category_id: string;
  amount: number;
  month: number;
  year: number;
}
